Documentation Cleanup - July - part 1 (#29802)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/authoring-builds/build_environment.adoc

.gradle.properties
----
org.gradle.java.home=/path/to/your/java/home
----
3. Environment Variable
+
[source,text]
----
$ export JAVA_HOME=/path/to/your/java/home
----

> == The `gradle.properties` file

Reviewer: I really like this new section as it really brings everything together. One thing that I think would make a great addition here is mentioning that the `gradle.properties` file can be located in `GRADLE_USER_HOME`, the root project directory, or `GRADLE_HOME`. It is sort of mentioned above in the _Priority for configurations_ section, but you could potentially go into a little more detail here.

Author: Added. Great suggestion.

[[sec:project_properties]]
== Project properties

Project properties are specific to your Gradle project, they can be used to customize your build.
Project properties can be accessed in your build files and get passed in from an external source when your build is executed.
Project properties can be retrieved lazily using `providers.gradleProperty()`.

=== Setting a project property

You have four options to add project properties, listed in order of priority:

1. *Command Line:* You can add project properties directly to your link:{groovyDslPath}/org.gradle.api.Project.html[Project] object via the `-P` command line option.
+
[source,text]
----
$ ./gradlew build -PmyProperty='Hi, world'
----

> 2. *System Property:* Gradle creates specially-named system properties for project properties which you can set using the `-D` command line flag or `gradle.properties` file. For the project property `myProperty`, the system property created is called `org.gradle.project.__prop__`.

Reviewer: ```suggestion
2. *System Property:* Gradle creates specially-named system properties for project properties which you can set using the `-D` command line flag or `gradle.properties` file. For the project property `myProperty`, the system property created is called `org.gradle.project.__myProperty__`.
```

+
[source,text]
----
$ ./gradlew build -Dorg.gradle.project.myProperty='Hi, world'
----
+
.gradle.properties
[source,properties]
----
org.gradle.project.myProperty='Hi, world'
----
3. *Gradle Properties File:* You can also set project properties in `gradle.properties` files.
+
.gradle.properties
[source,text]
----
myProperty='Hi, world'
----
4. *Environment Variables:* You can set project properties with environment variables. If the environment variable name looks like `ORG_GRADLE_PROJECT___prop__=somevalue`, then Gradle will set a `prop` property on your project object, with the value of `somevalue`.
+
[source,text]
----
$ export ORG_GRADLE_PROJECT_myProperty='Hi, world'
----
+
This feature is very useful when you don't have admin rights to a continuous integration server and need to set property values that are not easily visible.
Since you cannot use the `-P` option in that scenario nor change the system-level configuration files, the correct strategy is to change the configuration of your continuous integration build job, adding an environment variable setting that matches an expected pattern.

> This won't be visible to normal users on the system.

Reviewer: This example is strange because if you don't have admin rights on a CI server, and are unable to add `-P` to command line invocations, then you most likely don't have permission to add environment variables either. I would even say that typically admins are the _only_ ones that can define environment variables. For example, I can't set environment variables in [gradle/gradle](https://github.com/gradle/gradle), an admin has to do that, but I could create a PR to modify [a workflow file's command line invocation](https://github.com/gradle/gradle/blob/656ada7e12aad698cf1eb6f78e53ac48dcd01b8c/.github/workflows/contributor-pr.yml#L54) as a developer.

The sentiment that using environment variables for CI servers is good to have included in the docs, but I would drop the whole "admin CI server" example and just add something like this to the above list item:

> 4. *Environment Variables:* You can set project properties with [...]. This is typically the preferred way to supply project properties, especially secrets, to unattended builds like those running on CI servers.

Author: Sounds good. Added.

[[sec:project_properties]]
== Project properties

> Project properties are specific to your Gradle project, they are variables or configuration blocks defined in your build script or in the `Project` object.

Reviewer: This is a little strange to me because you don't define project properties in build scripts. They are defined using one of the 4 options shown below (command line, properties file, environment variables). Maybe "defined" is not the best word to use here, maybe "referenced"? 

Also, what does it mean "or configuration blocks" and "or in the `Project` object"?

Author: It is a poor choice of words, defined should be created. Rewording this.

Project properties can be retrieved:

- Lazily: `providers.gradleProperty()`

> - Eagerly: `project.ext()`, `project.findProperty()`, `project.property()`, `project.('<propertyName>')`

Reviewer: I think the docs should be exhaustive because users may see these in the wild and come to the docs looking for an explanation, but this really highlights one of the main complaints which is that there's just too many ways to do things.

Do we have a preferred way to reference properties? My guess would be lazily via `providers.gradleProperty()`. If you really need a value eagerly you can call `providers.gradleProperty(...).get()`.

It would be nice to just say that the lazy method is the recommended way but you might see these other ways in older projects.

If a project property is referenced but does not exist, an exception may be thrown and the build will fail.

> You should check for the existence of a project property before directly accessing it with `project.('<propertyName>')` by using the link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:hasProperty(java.lang.String)[`project.hasProperty()`] method.

Reviewer: If the `providers.gradleProperty()` method is the preferred way to access properties, then you can do all this using the `Provider` APIs. `providers.gradleProperty()` won't throw an exception if the property doesn't exist, only if `.get()` is called when there is no value present, but you can check for presence by using `.isPresent`.

Author: I went for a 180 and got too specific. Removing this,

=== Setting a project property

You have four options to add project properties, listed in order of priority:

1. *Command Line:* You can add project properties directly to your link:{groovyDslPath}/org.gradle.api.Project.html[Project] object via the `-P` command line option.
+
[source,text]
----
$ ./gradlew build -PmyProperty='Hi, world'
----

> 2. *As a System Property:* Gradle creates specially-named system properties for project properties which you can set using the `-D` command line flag or `gradle.properties` file. For the project property `myProperty`, the system property created is called `org.gradle.project.__prop__`.

Reviewer: This is more consistent with the other options.

```suggestion
2. *System Property:* Gradle creates specially-named system properties for project properties which you can set using the `-D` command line flag or `gradle.properties` file. For the project property `myProperty`, the system property created is called `org.gradle.project.__prop__`.
```

+
[source,text]
----
$ ./gradlew build -Dorg.gradle.project.myProperty='Hi, world'
----
+
.gradle.properties
[source,properties]
----
org.gradle.project.myProperty='Hi, world'
----
3. *Gradle Properties File:* You can also set project properties in `gradle.properties` files.
+
.gradle.properties
[source,text]
----
myProperty='Hi, world'
----

> 4. *Environment Variables:* You can set project properties with environment variables. If the environment variable name looks like `ORG_GRADLE_PROJECT___prop__=somevalue`, then Gradle will set a `prop` property on your project object, with the value of `somevalue`.

Reviewer: ```suggestion
4. *Environment Variables:* You can set project properties with environment variables. If the environment variable name looks like `ORG_GRADLE_PROJECT___myProperty__='Hi, world'`, then Gradle will set a `myProperty` property on your project object, with the value of `Hi, world`.
```

Reviewer: I love that you aligned all the examples, by the way! 👏

Reviewer: It looks like you missed adding this one. All other examples use `myProperty=Hi, world` except this one.

[source,kotlin]
.build.gradle.kts
----
tasks.register("gradlePropertyTest") {
    val isCI = providers.gradleProperty("org.gradle.caching.debug")
    doLast {
        println(isCI.get())
    }
}
----

And the `gradle.properties` file contains the following line:

[source,properties]
.gradle.properties
----
org.gradle.console=verbose
----

Then running the `gradlePropertyTest` task yields the following results:

[source,text]
----
$ ./gradlew gradlePropertyTest

> Task :buildSrc:generateExternalPluginSpecBuilders UP-TO-DATE
> Task :buildSrc:extractPrecompiledScriptPluginPlugins UP-TO-DATE
> Task :buildSrc:compilePluginsBlocks UP-TO-DATE
[...]
> Task :buildSrc:classes UP-TO-DATE
> Task :buildSrc:jar UP-TO-DATE
> Task :buildSrc:inspectClassesForKotlinIC UP-TO-DATE
> Task :model:gradlePropertyTest
verbose
----

The console output is verbose and therefore correct.
Let's see what happens when we run the same task with the added command line option (note that the `gradle.properties` file has not changed):

[source,text]
----
$ ./gradlew :gradlePropertyTest -Dorg.gradle.console=auto

> Task :model:gradlePropertyTest
verbose
----

As expected, the console output is short.

> However, the output of `providers.gradleProperty("org.gradle.caching.debug")` is incorrectly `verbose`.

Reviewer: I honestly don't follow this example at all. 😅 It seems like it's describing a very specific edge case.

If not referencing Gradle properties in build logic is our stance, I would just leave it at that and not go into a length example of why. If there are better APIs to use instead then this would be a good place to mention them. For example, the recently added [BuildFeatures](https://docs.gradle.org/current/javadoc/org/gradle/api/configuration/BuildFeatures.html) API. I'm not sure what our stance is on using [`Gradle#getStartParameter()`](https://docs.gradle.org/current/javadoc/org/gradle/api/invocation/Gradle.html#getStartParameter()), but that appears to be a way to get the value of many Gradle properties.

Author: Did the 180 thing there as well. Removing it.
